src/terralib.lua

```diff
@@ -3065,6 +3070,19 @@ function typecheck(topexp,luaenv,simultaneousdefinitions)
         local body = checkblock(s.body)
         return copyobject(s,{condition = e, body = body})
     end
+    local function checkexpintegral(re)
+        local e = checkexp(re)
+        if not e.type:isintegral() then
```

Comment:
Do we need to check that all the case expressions are (or coerce to?) the same type?

Reply:
LLVM doesn't care if the case expression is the exact same type, only that it resolves to some integral constant it can use to construct the jump statement. If we choose to expand the switch statement to accept other types, it might be worth putting in a more strict requirement, but since any coercion requirement should still allow the implicit conversion of integers to any comparable integer type I don't think it's necessary to implement it right now.